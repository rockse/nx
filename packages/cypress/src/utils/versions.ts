export const nxVersion = require('../../package.json').version;
export const eslintPluginCypressVersion = '^2.10.3';
export const typesNodeVersion = '16.11.7';
export const cypressVersion = '^10.1.0';
export const cypressWebpackVersion = '^2.0.0';
export const webpackHttpPluginVersion = '^5.5.0';
