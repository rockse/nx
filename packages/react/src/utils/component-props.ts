import * as ts from 'typescript';
import { getComponentPropsInterface } from './ast-utils';

// TODO: candidate to refactor with the angular component story
export function getArgsDefaultValue(property: ts.SyntaxKind): string {
  const typeNameToDefault: Record<number, any> = {
    [ts.SyntaxKind.StringKeyword]: "''",
    [ts.SyntaxKind.NumberKeyword]: 0,
    [ts.SyntaxKind.BooleanKeyword]: false,
  };

  const resolvedValue = typeNameToDefault[property];
  if (typeof resolvedValue === undefined) {
    return "''";
  } else {
    return resolvedValue;
  }
}

export function getDefaultsForComponent(
  sourceFile: ts.SourceFile,
  cmpDeclaration: ts.Node
) {
  const propsInterface = getComponentPropsInterface(sourceFile, cmpDeclaration);

  let propsTypeName: string = null;
  let props: {
    name: string;
    defaultValue: any;
  }[] = [];
  let argTypes: {
    name: string;
    type: string;
    actionText: string;
  }[] = [];

  if (propsInterface) {
    propsTypeName = propsInterface.name.text;
    props = propsInterface.members.map((member: ts.PropertySignature) => {
      if (member.type.kind === ts.SyntaxKind.FunctionType) {
        argTypes.push({
          name: (member.name as ts.Identifier).text,
          type: 'action',
          actionText: `${(member.name as ts.Identifier).text} executed!`,
        });
      } else {
        return {
          name: (member.name as ts.Identifier).text,
          defaultValue: getArgsDefaultValue(member.type.kind),
        };
      }
    });
    props = props.filter((p) => p && p.defaultValue !== undefined);
  }
  return { propsTypeName, props, argTypes };
}

export function getImportForType(sourceFile: ts.SourceFile, typeName: string) {
  return sourceFile.statements.find(
    (statement: ts.Node) =>
      ts.isImportDeclaration(statement) &&
      statement.getText().includes(typeName)
  );
}
