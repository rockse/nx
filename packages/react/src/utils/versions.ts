export const nxVersion = require('../../package.json').version;

export const reactVersion = '18.2.0';
export const reactDomVersion = '18.2.0';
export const reactIsVersion = '18.2.0';
export const typesReactVersion = '18.0.13';
export const typesReactDomVersion = '18.0.5';
export const typesReactIsVersion = '17.0.3';

export const typesNodeVersion = '18.0.0';

export const styledComponentsVersion = '5.3.5';
export const typesStyledComponentsVersion = '5.1.25';

export const emotionStyledVersion = '11.9.3';
export const emotionReactVersion = '11.9.3';
export const emotionBabelPlugin = '11.9.2';

// WARNING: This needs to be in sync with Next.js' dependency or else there might be issues.
export const styledJsxVersion = '5.0.2';

export const reactRouterDomVersion = '6.3.0';
export const typesReactRouterDomVersion = '5.3.3';

export const testingLibraryReactVersion = '13.3.0';

export const reduxjsToolkitVersion = '1.8.2';
export const reactReduxVersion = '8.0.2';
export const reactTestRendererVersion = '18.2.0';

export const eslintPluginImportVersion = '2.26.0';
export const eslintPluginJsxA11yVersion = '6.5.1';
export const eslintPluginReactVersion = '7.30.0';
export const eslintPluginReactHooksVersion = '4.6.0';

export const babelPluginStyledComponentsVersion = '1.10.7';

// used for cypress component tests
export const cypressReactVersion = '^6.0.0';
