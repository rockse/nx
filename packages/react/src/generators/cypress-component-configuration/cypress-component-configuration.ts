import { cypressComponentProject } from '@nrwl/cypress';
import {
  addDependenciesToPackageJson,
  generateFiles,
  joinPathFragments,
  ProjectConfiguration,
  readProjectConfiguration,
  Tree,
  updateJson,
  visitNotIgnoredFiles,
} from '@nrwl/devkit';
import * as ts from 'typescript';
import { getComponentNode } from '../../utils/ast-utils';
import { cypressReactVersion } from '../../utils/versions';
import componentTestGenerator from '../component-test/component-test';
import { CypressComponentConfigurationSchema } from './schema';

const allowedFileExt = new RegExp(/\.[jt]sx?/g);
const isSpecFile = new RegExp(/(spec|test)\./g);

/**
 * This is for using cypresses own Component testing, if you want to use test
 * storybook components then use componentCypressGenerator instead.
 *
 */
export async function cypressComponentConfigGenerator(
  tree: Tree,
  options: CypressComponentConfigurationSchema
) {
  const projectConfig = readProjectConfiguration(tree, options.project);
  const baseDepInstallTask = await cypressComponentProject(tree, {
    project: options.project,
    skipFormat: true,
  });
  const installReactDeps = addDeps(tree);

  addFiles(tree, projectConfig, options);
  updateTsConfig(tree, projectConfig);

  return () => {
    baseDepInstallTask();
    installReactDeps();
  };
}

function addDeps(tree: Tree) {
  const devDeps = {
    '@cypress/react': cypressReactVersion,
  };
  // TODO(caleb): add swc deps if using swc as a compiler?
  return addDependenciesToPackageJson(tree, {}, devDeps);
}

function addFiles(
  tree: Tree,
  projectConfig: ProjectConfiguration,
  options: CypressComponentConfigurationSchema
) {
  const cypressConfigPath = joinPathFragments(
    projectConfig.root,
    'cypress.config.ts'
  );
  if (tree.exists(cypressConfigPath)) {
    tree.delete(cypressConfigPath);
  }

  generateFiles(
    tree,
    joinPathFragments(__dirname, 'files'),
    projectConfig.root,
    {
      tpl: '',
    }
  );

  if (options.generateTests) {
    visitNotIgnoredFiles(tree, projectConfig.sourceRoot, (filePath) => {
      if (isComponent(tree, filePath)) {
        componentTestGenerator(tree, {
          project: options.project,
          componentPath: filePath,
        });
      }
    });
  }
}

function updateTsConfig(tree: Tree, projectConfig: ProjectConfiguration) {
  const tsConfigPath = joinPathFragments(
    projectConfig.root,
    projectConfig.projectType === 'library'
      ? 'tsconfig.lib.json'
      : 'tsconfig.app.json'
  );
  if (tree.exists(tsConfigPath)) {
    updateJson(tree, tsConfigPath, (json) => {
      const excluded = new Set([
        ...(json.exclude || []),
        'cypress/**/*',
        'cypress.config.ts',
        '**/*.cy.ts',
        '**/*.cy.js',
        '**/*.cy.tsx',
        '**/*.cy.jsx',
      ]);

      json.exclude = Array.from(excluded);
      return json;
    });
  }

  const projectBaseTsConfig = joinPathFragments(
    projectConfig.root,
    'tsconfig.json'
  );
  if (tree.exists(projectBaseTsConfig)) {
    updateJson(tree, projectBaseTsConfig, (json) => {
      if (json.references) {
        const hasCyTsConfig = json.references.some(
          (r) => r.path === './tsconfig.cy.json'
        );
        if (!hasCyTsConfig) {
          json.references.push({ path: './tsconfig.cy.json' });
        }
      } else {
        const excluded = new Set([
          ...(json.exclude || []),
          'cypress/**/*',
          'cypress.config.ts',
          '**/*.cy.ts',
          '**/*.cy.js',
          '**/*.cy.tsx',
          '**/*.cy.jsx',
        ]);

        json.exclude = Array.from(excluded);
      }
      return json;
    });
  }
}

function isComponent(tree: Tree, filePath: string): boolean {
  if (isSpecFile.test(filePath) || !allowedFileExt.test(filePath)) {
    return false;
  }

  const content = tree.read(filePath, 'utf-8');
  const sourceFile = ts.createSourceFile(
    filePath,
    content,
    ts.ScriptTarget.Latest,
    true
  );

  const cmpDeclaration = getComponentNode(sourceFile);
  return !!cmpDeclaration;
}

export default cypressComponentConfigGenerator;
