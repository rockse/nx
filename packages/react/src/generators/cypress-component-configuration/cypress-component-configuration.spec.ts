import { assertMinimumCypressVersion } from '@nrwl/cypress/src/utils/cypress-version';
import { readJson, Tree } from '@nrwl/devkit';
import { createTreeWithEmptyWorkspace } from '@nrwl/devkit/testing';
import { Linter } from '@nrwl/linter';
import componentGenerator from '../component/component';
import libraryGenerator from '../library/library';
import { cypressComponentConfigGenerator } from './cypress-component-configuration';

jest.mock('@nrwl/cypress/src/utils/cypress-version');
describe('React:CypressComponentTestConfiguration', () => {
  let tree: Tree;
  let mockedAssertCypressVersion: jest.Mock<
    ReturnType<typeof assertMinimumCypressVersion>
  > = assertMinimumCypressVersion as never;
  beforeEach(() => {
    tree = createTreeWithEmptyWorkspace();
  });
  it('should generate cypress component test config', async () => {
    mockedAssertCypressVersion.mockReturnValue();

    await libraryGenerator(tree, {
      linter: Linter.EsLint,
      name: 'some-lib',
      skipFormat: true,
      skipTsConfig: false,
      style: 'scss',
      unitTestRunner: 'none',
      component: true,
    });

    await cypressComponentConfigGenerator(tree, {
      project: 'some-lib',
      generateTests: false,
    });

    const config = tree.read('libs/some-lib/cypress.config.ts', 'utf-8');
    expect(config).toContain(
      "import { nxComponentTestingPreset } from '@nrwl/react/plugins/component-testing"
    );
    expect(config).toContain('component: nxComponentTestingPreset(__dirname),');

    const cyTsConfig = readJson(tree, 'libs/some-lib/tsconfig.cy.json');
    expect(cyTsConfig.include).toEqual([
      'cypress.config.ts',
      '**/*.cy.ts',
      '**/*.cy.tsx',
      '**/*.cy.js',
      '**/*.cy.jsx',
      '**/*.d.ts',
    ]);
    const libTsConfig = readJson(tree, 'libs/some-lib/tsconfig.lib.json');
    expect(libTsConfig.exclude).toEqual(
      expect.arrayContaining([
        'cypress/**/*',
        'cypress.config.ts',
        '**/*.cy.ts',
        '**/*.cy.js',
        '**/*.cy.tsx',
        '**/*.cy.jsx',
      ])
    );
    const baseTsConfig = readJson(tree, 'libs/some-lib/tsconfig.json');
    expect(baseTsConfig.references).toEqual(
      expect.arrayContaining([{ path: './tsconfig.cy.json' }])
    );
  });

  it('should generate tests for existing tsx components', async () => {
    mockedAssertCypressVersion.mockReturnValue();

    await libraryGenerator(tree, {
      linter: Linter.EsLint,
      name: 'some-lib',
      skipFormat: true,
      skipTsConfig: false,
      style: 'scss',
      unitTestRunner: 'jest',
      component: true,
    });
    await componentGenerator(tree, {
      name: 'another-cmp',
      project: 'some-lib',
      style: 'scss',
    });

    await cypressComponentConfigGenerator(tree, {
      project: 'some-lib',
      generateTests: true,
    });

    expect(tree.exists('libs/some-lib/src/lib/some-lib.cy.tsx')).toBeTruthy();
    const compTest = tree.read(
      'libs/some-lib/src/lib/some-lib.cy.tsx',
      'utf-8'
    );
    expect(compTest).toMatchSnapshot();
    expect(tree.exists('libs/some-lib/src/lib/some-lib.cy.tsx')).toBeTruthy();
    const compTestNested = tree.read(
      'libs/some-lib/src/lib/another-cmp/another-cmp.cy.tsx',
      'utf-8'
    );
    expect(compTestNested).toMatchSnapshot();
    expect(
      tree.exists('libs/some-lib/src/lib/another-cmp/another-cmp.spec.cy.tsx')
    ).toBeFalsy();
  });
  it('should generate tests for existing js components', async () => {
    mockedAssertCypressVersion.mockReturnValue();

    await libraryGenerator(tree, {
      linter: Linter.EsLint,
      name: 'some-lib',
      skipFormat: true,
      skipTsConfig: false,
      style: 'scss',
      unitTestRunner: 'jest',
      js: true,
    });
    await componentGenerator(tree, {
      name: 'some-cmp',
      flat: true,
      project: 'some-lib',
      style: 'scss',
      js: true,
    });
    await componentGenerator(tree, {
      name: 'another-cmp',
      project: 'some-lib',
      style: 'scss',
      js: true,
    });

    await cypressComponentConfigGenerator(tree, {
      project: 'some-lib',
      generateTests: true,
    });

    expect(tree.exists('libs/some-lib/src/lib/some-cmp.cy.js')).toBeTruthy();
    const compTest = tree.read('libs/some-lib/src/lib/some-cmp.cy.js', 'utf-8');
    expect(compTest).toMatchSnapshot();
    expect(
      tree.exists('libs/some-lib/src/lib/another-cmp/another-cmp.cy.js')
    ).toBeTruthy();
    const compTestNested = tree.read(
      'libs/some-lib/src/lib/another-cmp/another-cmp.cy.js',
      'utf-8'
    );
    expect(compTestNested).toMatchSnapshot();
    expect(
      tree.exists('libs/some-lib/src/lib/another-cmp/another-cmp.spec.cy.js')
    ).toBeFalsy();
  });
});
