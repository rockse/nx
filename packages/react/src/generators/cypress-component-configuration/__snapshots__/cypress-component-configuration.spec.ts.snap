// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`React:CypressComponentTestConfiguration should generate tests for existing js components 1`] = `
"import * as React from 'react'
import { mount } from '@cypress/react'
import SomeCmp from './some-cmp'


describe(SomeCmp.name, () => {
  it('renders', () => {
    mount(<SomeCmp />)
  })
})

"
`;

exports[`React:CypressComponentTestConfiguration should generate tests for existing js components 2`] = `
"import * as React from 'react'
import { mount } from '@cypress/react'
import AnotherCmp from './another-cmp'


describe(AnotherCmp.name, () => {
  it('renders', () => {
    mount(<AnotherCmp />)
  })
})

"
`;

exports[`React:CypressComponentTestConfiguration should generate tests for existing tsx components 1`] = `
"import * as React from 'react'
import { mount } from '@cypress/react'
import SomeLib, { SomeLibProps } from './some-lib'


describe(SomeLib.name, () => {
  let props: SomeLibProps;

  beforeEach(() => {
    props = {
    }
  })

  it('renders', () => {
    mount(<SomeLib {...props}/>)
  })
})

"
`;

exports[`React:CypressComponentTestConfiguration should generate tests for existing tsx components 2`] = `
"import * as React from 'react'
import { mount } from '@cypress/react'
import AnotherCmp, { AnotherCmpProps } from './another-cmp'


describe(AnotherCmp.name, () => {
  let props: AnotherCmpProps;

  beforeEach(() => {
    props = {
    }
  })

  it('renders', () => {
    mount(<AnotherCmp {...props}/>)
  })
})

"
`;
